### PR TITLE
feat: Parse templates in _answers_file setting

### LINKED ISSUE #866: Dynamic answer file naming not working
### Discussed in https://github.com/copier-org/copier/discussions/860

<div type='discussions-op-text'>

<sup>Originally posted by **nilsdebruin** November  3, 2022</sup>
Hi all!

It is possible to define an answer file in the copier.yml like for example:

`_answers_file: .copier-answers-copier-template.yml`

It would be really nice if this could be extended with a dynamic value that you could set as a jinja variable like this:

`_answers_file: .copier-answers-copier-template-{{module_name}}.yml`

Where module_name is an answer you can give in copier.yml like:

```
module_name:
  type: str
```

This would make It possible to create multiple instances of a template in the same structure like:

```
root_level/{{module_name_1}}
root_level/{{module_name_2}}
```

This would create two answer files:

```
.copier-answers-copier-template-module_name_1.yml
.copier-answers-copier-template-module_name_2.yml
```

These template instances could then be updated based on these answer files.</div>

### PR DESCRIPTION
**Pull request by icj217**
Adds the ability for `_answers_file` to contain template variables. For example:
```yaml
# copier.yml
_answers_file: ".copier-answers.{{ foo }}.yml"

foo:
  type: str
```

Closes #866

Since #882 seems to have been abandoned and the tests weren't actually working, I've opened up this PR instead.

### CODE REVIEW

**Comment by nilsdebruin** [comment]
Great! Would be nice if this feature is added!

File: copier/main.py [hunk 1]
```diff
@@ -417,7 +417,9 @@ def answers_relpath(self) -> Path:
         2. Template default.
         3. Copier default.
         """
-        return self.answers_file or self.template.answers_relpath
+        path = self.answers_file or self.template.answers_relpath
+        template = self.jinja_env.from_string(str(path))
+        return Path(template.render(**self.answers.combined))
```
Comment (sisp) [inline copier/main.py:422]: I think you can simplify this part by using [`_render_string()`](https://github.com/copier-org/copier/blob/beb174f64611d1343621ca5b8b5a7d668eee94ab/copier/main.py#L566,L574):

```suggestion
        return Path(self._render_string(str(path)))
```

File: tests/test_answersfile_templating.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        defaults=True,
+        overwrite=True,
+    )
+    first_answers_file = (
+        ".copier-answers-mymodule.yml"
+        if answers_file is None
+        else ".changed-by-user.yml"
+    )
+    path = Path(tmp_path) / (first_answers_file)
+    assert path.exists()
```
Comment (sisp) [inline tests/test_answersfile_templating.py:56]: `path` is already a `pathlib.Path` object. I would simplify this a bit:

```suggestion
    assert (tmp_path / first_answers_file).exists()
```

File: tests/test_answersfile_templating.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        tmp_path,
+        {"module_name": "anothermodule"},
+        defaults=True,
+        overwrite=True,
+    )
+
+    # Assert second one created
+    second_answers_file = ".copier-answers-anothermodule.yml"
+    path = Path(tmp_path) / (second_answers_file)
+    assert path.exists()
```
Comment (sisp) [inline tests/test_answersfile_templating.py:71]: `path` is already a `pathlib.Path` object. I would simplify this a bit:

```suggestion
    assert (tmp_path / second_answers_file).exists()
```

**Review by sisp — CHANGES_REQUESTED** [review]
I have a couple of suggestions for improvement. But overall it looks good to me. 🙂

File: copier/main.py (hunk 1 above)
Comment (icj217) [reply copier/main.py:422]: I tried this out but got a recursion error in my tests:
```
tests/test_answersfile_templating.py:42: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
copier/main.py:927: in run_auto
    return run_copy(src_path, dst_path, data, **kwargs)
copier/main.py:890: in run_copy
    with Worker(src_path=src_path, dst_path=Path(dst_path), **kwargs) as worker:
copier/main.py:184: in __exit__
    raise value
copier/main.py:891: in run_copy
    worker.run_copy()
copier/main.py:640: in run_copy
    src_abspath = self.template_copy_root
/usr/local/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/functools.py:1001: in __get__
    val = self.func(instance)
copier/main.py:613: in template_copy_root
    subdir = self._render_string(self.template.subdirectory) or ""
copier/main.py:587: in _render_string
    return tpl.render(**self._render_context())
copier/main.py:242: in _render_context
    "answers_file": self.answers_relpath,
/usr/local/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/functools.py:1001: in __get__
    val = self.func(instance)
copier/main.py:421: in answers_relpath
    return Path(self._render_string(str(path)))
copier/main.py:587: in _render_string
    return tpl.render(**self._render_context())
copier/main.py:242: in _render_context
    "answers_file": self.answers_relpath,
E   RecursionError: maximum recursion depth exceeded while calling a Python object
```

Comment (sisp) [reply copier/main.py:422]: Ah, yes, I see. Ignore my suggestion then.

**Comment by icj217** [comment]
@yajo Would you mind taking a look and approving the workflow runs?

**Comment by icj217** [comment]
@sisp Are you able to trigger the workflow runs for this PR?

**Comment by sisp** [comment]
I don't think I can help because I have no permissions for this project.

**Comment by icj217** [comment]
@yajo Sorry for the bump but wondering if you can approve these workflow runs?

**Comment by yajo** [comment]
Triggered! No concerns, just I had 120 emails from copier 😂

**Review by yajo — APPROVED** [review]
Pretty clean, thanks!

**Pull request merged**